### PR TITLE
Disallow non-utf8 chars as input to phonenumbermatcher API

### PR DESCRIPTION
**Pull request by penmetsaa**
- Matcher.HasNext() returns false even in the first run when the input has non-utf8 chars in it.
-  Done by flagging such inputs when creating phonenumbermatcher obj.
- More details in b/201576863

### CODE REVIEW

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (lararennie) [inline cpp/src/phonenumbers/phonenumbermatcher.cc:630]: For style reasons, we don't use exceptions in C++.

Comment (penmetsaa) [reply cpp/src/phonenumbers/phonenumbermatcher.cc]: Thanks Lara..
Sorry, earlier went ahead with the suggestion as we mistakenly assumed that we are not able to replicate internally and something to do with the flags like "-fsanitize=address,fuzzer-no-link" as we see lot memory leak warnings when run that way. We were able to reproduce the issue internally as well. cl/416205516

I see we hit this issue only when the input is large (> 2000 chars) and the index going negative  in next iteration. I updated the code to check the range of index. Will debug more and see the root cause and have proper fix (like having a cap on the input length / using unsigned int for index...

Comment (penmetsaa) [reply cpp/src/phonenumbers/phonenumbermatcher.cc]: Roughly till length 3500 we are not hitting the issue. As I am not familiar with CPP, could not debug why the index becomes negative. Maybe we can go ahead with the present solution?

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (lararennie) [inline cpp/src/phonenumbers/phonenumbermatcher.cc:659]: I don't think this is the right place to check this. That means you do this check every time you call next.
Can we do it in the constructor? The text does not change. And store it as a boolean, and if it's false, return false from HasNext?

Comment (penmetsaa) [reply cpp/src/phonenumbers/phonenumbermatcher.cc:659]: Hi Lara,

I have not blocked PhoneNumberMatcher creation as issues can be returned only through returning exceptions in constructor. As adding CPP exceptions are not encouraged (also introduce backward incompatibility -compile errors), lazily returning no matches (in Find() API), similar to earlier commits.

Yes, I would have explored alternate options like:
- doing this check in constructor and flag it to reference later during Find()/any as mentioned in https://stackoverflow.com/a/45643989
- I do not see other options that are innocuous. Please suggest.

File: cpp/test/phonenumbers/phonenumbermatcher_test.cc
Comment (tvislavski) [inline cpp/test/phonenumbers/phonenumbermatcher_test.cc:1097]: EXPECT_TRUE(true)? :)

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (katbohm) [reply cpp/src/phonenumbers/phonenumbermatcher.cc:659]: Doing this check in constructor & adding a boolean flag sounds reasonable to me, to avoid checking the entire input text repeatedly.

File: cpp/test/phonenumbers/phonenumbermatcher_test.cc
Comment (penmetsaa) [reply cpp/test/phonenumbers/phonenumbermatcher_test.cc]: My bad. UT is not updated, checked only through debug results.
Updated now.

File: cpp/test/phonenumbers/phonenumbermatcher_test.cc
Comment (tvislavski) [inline cpp/test/phonenumbers/phonenumbermatcher_test.cc:1101]: Can't you say here simply `EXPECT_FALSE(matcher->HasNext())`? Then you can omit `while` loop and counter

Comment (penmetsaa) [reply cpp/test/phonenumbers/phonenumbermatcher_test.cc]: +1. My bad.
I do not see reason now why I though it would be clear to mention that no tries will happen. :)

**Review by tvislavski — APPROVED** [review]
LGTM, but please wait for final approval from Lara or Katerina

**Review by katbohm — COMMENTED** [review]
Do we document somewhere that the input for the matcher needs to be in clean utf8 and otherwise matcher.HasNext() returns false?

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (penmetsaa) [reply cpp/src/phonenumbers/phonenumbermatcher.cc]: Done. Updated relevant .h file. Please verify whether it fine..

**Comment by penmetsaa** [comment]
> 

Done. Updated relevant .h file. Please verify whether it fine..

**Review by katbohm — DISMISSED** [review]
LGTM

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (lararennie) [reply cpp/src/phonenumbers/phonenumbermatcher.cc]: Indeed :)

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (lararennie) [inline cpp/src/phonenumbers/phonenumbermatcher.cc:1]: Why did you drop the copyright?

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (lararennie) [inline cpp/src/phonenumbers/phonenumbermatcher.cc:410]: clean -> valid?

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (lararennie) [inline cpp/src/phonenumbers/phonenumbermatcher.cc:430]: Why the extra blank line?

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (lararennie) [reply cpp/src/phonenumbers/phonenumbermatcher.cc:410]: Also var names in c++ should be underscores, not camel case, so is_input_in_clean_utf8_

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (lararennie) [inline cpp/src/phonenumbers/phonenumbermatcher.cc:435]: Better method name is just IsInputUtf8 (rather than "or not" at the end)

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (lararennie) [inline cpp/src/phonenumbers/phonenumbermatcher.cc:439]: better: return number_as_unicode.UTF8WasValid(); instead of the if/else?

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (lararennie) [inline cpp/src/phonenumbers/phonenumbermatcher.cc:434]: How/why is this static, doesn't it operate on "text_" which is a local var?

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (lararennie) [inline cpp/src/phonenumbers/phonenumbermatcher.cc:436]: Comment would be better if it explained *why* -> but then again, I wouldn't put it here, I would put it where you use the variable. Here the code doesn't need a comment IMO.

File: cpp/test/phonenumbers/phonenumbermatcher_test.cc
Comment (lararennie) [inline cpp/test/phonenumbers/phonenumbermatcher_test.cc:1093]: Maybe a better comment: "Since the input text contains invalid UTF-8, we do not return any matches."

File: cpp/src/phonenumbers/phonenumbermatcher.h
Comment (lararennie) [inline cpp/src/phonenumbers/phonenumbermatcher.h:105]: Remove "API to" and start with just "Checks if the ...."

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (lararennie) [reply cpp/src/phonenumbers/phonenumbermatcher.cc:410]: so perhaps: is_input_valid_utf8

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (lararennie) [inline cpp/src/phonenumbers/phonenumbermatcher.cc:644]: This is where I would put a comment explaining *why* we early exit if it is not valid utf8 :)

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (penmetsaa) [reply cpp/src/phonenumbers/phonenumbermatcher.cc:439]: +1. Done.

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (penmetsaa) [reply cpp/src/phonenumbers/phonenumbermatcher.cc:435]: Done. +1

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (penmetsaa) [reply cpp/src/phonenumbers/phonenumbermatcher.cc:430]: Removed.

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (penmetsaa) [reply cpp/src/phonenumbers/phonenumbermatcher.cc:410]: Done. +1

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (penmetsaa) [reply cpp/src/phonenumbers/phonenumbermatcher.cc:1]: My bad. Did by mistake.

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (penmetsaa) [reply cpp/src/phonenumbers/phonenumbermatcher.cc:436]: Done. Thanks Lara :)

File: cpp/test/phonenumbers/phonenumbermatcher_test.cc
Comment (penmetsaa) [reply cpp/test/phonenumbers/phonenumbermatcher_test.cc:1093]: Done.

File: cpp/src/phonenumbers/phonenumbermatcher.h
Comment (penmetsaa) [reply cpp/src/phonenumbers/phonenumbermatcher.h:105]: Done. Thanks Lara :)

File: cpp/src/phonenumbers/phonenumbermatcher.cc
Comment (penmetsaa) [reply cpp/src/phonenumbers/phonenumbermatcher.cc:644]: Done. Thanks Lara :)

File: cpp/src/phonenumbers/phonenumbermatcher.h
Comment (lararennie) [reply cpp/src/phonenumbers/phonenumbermatcher.h:105]: This one looks not done

**Review by lararennie — APPROVED** [review]

**Comment by lararennie** [comment]
Update the PR description though too, it is a bit confusing right now.

**Comment by penmetsaa** [comment]
> Update the PR description though too, it is a bit confusing right now.

Done. Thanks Lara :)

**Pull request merged**
